src/klement.jl

```diff
@@ -29,33 +29,77 @@ function SciMLBase.solve(prob::NonlinearProblem,
     xₙ = x
     xₙ₋₁ = x
     fₙ₋₁ = fₙ
-    for _ in 1:maxiters
-        xₙ = xₙ₋₁ - inv(J) * fₙ₋₁
-        fₙ = f(xₙ)
-        Δxₙ = xₙ - xₙ₋₁
-        Δfₙ = fₙ - fₙ₋₁
 
-        # Prevent division by 0
-        denominator = max.(J' .^ 2 * Δxₙ .^ 2, 1e-9)
+    # x is scalar
+    if x isa Number
+        J = 1.0
+        for _ in 1:maxiters
+            xₙ = xₙ₋₁ - fₙ₋₁ / J
+            fₙ = f(xₙ)
 
-        k = (Δfₙ - J * Δxₙ) ./ denominator
-        J += (k * Δxₙ' .* J) * J
+            iszero(fₙ) &&
+                return SciMLBase.build_solution(prob, alg, xₙ, fₙ;
+                                                retcode = ReturnCode.Success)
 
-        # Prevent inverting singular matrix
-        if det(J) ≈ 0
-            J = ArrayInterfaceCore.zeromatrix(x) + I
+            if isapprox(xₙ, xₙ₋₁, atol = atol, rtol = rtol)
+                return SciMLBase.build_solution(prob, alg, xₙ, fₙ;
+                                                retcode = ReturnCode.Success)
+            end
+
+            Δxₙ = xₙ - xₙ₋₁
+            Δfₙ = fₙ - fₙ₋₁
+
+            # Prevent division by 0
+            denominator = max(J^2 * Δxₙ^2, 1e-9)
+
+            k = (Δfₙ - J * Δxₙ) / denominator
+            J += (k * Δxₙ * J) * J
+
+            # Singularity test
+            if J < singular_tol
+                J = 1.0
+            end
+
+            xₙ₋₁ = xₙ
+            fₙ₋₁ = fₙ
         end
+        # x is a vector
+    else
+        J = init_J(x)
+        F = lu(J, check = false)
+        for _ in 1:maxiters
+            tmp = F \ fₙ₋₁
+            xₙ = xₙ₋₁ - tmp
+            fₙ = f(xₙ)
+
+            iszero(fₙ) &&
+                return SciMLBase.build_solution(prob, alg, xₙ, fₙ;
+                                                retcode = ReturnCode.Success)
+
+            if isapprox(xₙ, xₙ₋₁, atol = atol, rtol = rtol)
+                return SciMLBase.build_solution(prob, alg, xₙ, fₙ;
+                                                retcode = ReturnCode.Success)
+            end
+
+            Δxₙ = xₙ - xₙ₋₁
+            Δfₙ = fₙ - fₙ₋₁
+
+            # Prevent division by 0
+            denominator = max.(J' .^ 2 * Δxₙ .^ 2, 1e-9)
+
+            k = (Δfₙ - J * Δxₙ) ./ denominator
+            J += (k * Δxₙ' .* J) * J
+            F = lu(J, check = false)
 
-        iszero(fₙ) &&
-            return SciMLBase.build_solution(prob, alg, xₙ, fₙ;
-                                            retcode = ReturnCode.Success)
+            # Singularity test
+            if any(abs.(F.U[diagind(F.U)]) .< singular_tol)
```

Comment:
This allocate unnecessarily.

Reply:
I didn't get ```abs(F.factors[end, end])``` to work for static arrays (```F.factors``` gives error)

Reply:
And for non-static arrays ```F.factors[end, end]``` works fine if ```J``` isn't singular.
But if ```J``` is singular, then ```F.factors[end, end]``` didn't give the value that I excpected e.g. ```F.factors[end, end]=3.0``` but ```F.factors[end-1, end-1]=0.0```.

Reply:
And for non-static arrays ```abs(F.factors[end, end])``` worked fine if ```J``` was non-singular.
But it didn't work as I excpected when ```J``` was singular, i.e. it gave ```abs(F.factors[end, end])=3.0``` and  ```abs(F.factors[end-1, end-1])=0.0```, resulting in the solver ending up with ```NaN``` as output.

Reply:
Let's open up an issue on this. It may need to be specialized more, but for now we can go with this.